### ISSUE TITLE
Update WAGMI, viem and RainbowKit

### ISSUE THREAD

**Issue by H34D** [masa-finance/masa-react#590]
The above libs are outdated and need to be updated to support newer wallets like phantom

**Comment by juanmanso** [comment]
## Versioning breakdown

|  | _Current_ | _Latest_ |
| :--: | :--: | :---: |
| `wagmi` | ^1.4.13 | [2.12.1](https://www.npmjs.com/package/wagmi) |
| `viem` | ^1.21.4 | [2.18.5](https://www.npmjs.com/package/viem) |
| `rainbowkit` | ^1.3.7 | [2.1.3](https://www.npmjs.com/package/@rainbow-me/rainbowkit)  |

**Comment by juanmanso** [comment]
Moved this back to Ready since I'm working on the higher priority bug

**Comment by juanmanso** [comment]
Resuming work. Gonna inspect #514 for ideas on the implementation of the update

**Comment by juanmanso** [comment]
🟢 **[UPDATE]**

While reviewing the PR changes were way too many and hard to follow since there's just a single commit for the whole set of changes.

Instead what I ended up doing is a more hands-on approach where I start applying the changes myself to the latest stage of the repo. Still tinkering a bit but leaving an update of what I'm doing to keep communication fluid 💪

**Comment by juanmanso** [comment]
### Upgrading rainbowkit

Upgrading the `@rainbow-me/rainbowkit` dependency to its latest version throws the following warnings:

| ![Image](https://github.com/user-attachments/assets/298bcb63-65d4-4630-abb0-0f4c3d8b3bdf) |
| :--: |

This is coherent with their `package.json`'s peerDependencies object, where [viem needs to be 2.x](https://github.com/rainbow-me/rainbowkit/blob/048ab77d0f4fd6d1ea5b481a1ce618af6fd84f9f/packages/rainbowkit/package.json#L48) and [wagmi ^2.9.0](https://github.com/rainbow-me/rainbowkit/blob/048ab77d0f4fd6d1ea5b481a1ce618af6fd84f9f/packages/rainbowkit/package.json#L49).

An unexpected peerDependency appeared with `@celo/rainbowkit-celo`. However, they seem to not have updated their code as well since [their rainbowkit dependency is still set to 1.x](https://github.com/celo-org/rainbowkit-celo/blob/d38776594fdbad54da8ff243a6f4e4eeb50b4ced/packages/rainbowkit-celo/package.json#L75). We need to dig deeper into this but will be dismissed since it's not the focus of this ticket

**Comment by juanmanso** [comment]
After running

```sh
yarn upgrade @rainbow-me/rainbowkit@2 wagmi@2 viem@2.x
```

We got all the latest versions as shown on [this message](https://github.com/masa-finance/masa-react/issues/590#issuecomment-2258550814) with a few minor version's offset:

|  | _Current_ | _Latest_ |
| :--: | :--: | :---: |
| `wagmi` | ^1.4.13 | [2.12.2](https://www.npmjs.com/package/wagmi) |
| `viem` | ^1.21.4 | [2.18.7](https://www.npmjs.com/package/viem) |
| `rainbowkit` | ^1.3.7 | [2.1.3](https://www.npmjs.com/package/@rainbow-me/rainbowkit)  |

**Comment by juanmanso** [comment]
By looking at `viem`'s migration guide, it seems we have no breaking changes from its major version bump directly:

- https://viem.sh/docs/migration-guide.html#_2-x-x-breaking-changes

> [!NOTE]
> _This is mostly because we are not using `viem` directly but rather as a peer dependency for wagmi and rainbowkit_

**Comment by juanmanso** [comment]
**HOW TO DRILL DOWN THE CHANGES**

The repo has the whole e2e flow by integrating stories that would act like the end user of the library (in our case, the SBT app).

So starting from the end user, how do changes affect the pipeline?

### Stories

As on the SBT app, they import 2 key elements:
- `MasaProvider`. It consists of 2 providers:
  - `MasaWalletProvider`
  - `MasaStateProvider` 
- Custom hooks that interact with the provider. What we care most in this scope are:
  - `useWallet`
  - `useNetwork`

### MasaProvider

One important thing to mention is that `MasaStateProvider` makes use of TanStack/Query to manage the state and now it is a requirement for wagmi. Thus, since we cannot initiate 2 separated providers of TanStack/Query, we need to merge them.

The proposed solution on #514 was to just move it to:
- Move `QueryProvider` to the scope of `MasaWalletProvider`
- Create a wrapper provider for `MasaWalletProvider` that initiates:
 - `WagmiProvider`
 - `QueryProvider`
 - `RainbowkitProvider`
- Leave only Client management logic to `MasaStateProvider` (via `MasaClientProvider`)

Further detail of the changes will be given later 💪 


### Custom hooks

From a high-level perspective, it seems that only syntax changes are required and minor tweaks such as how internal functions/hooks are used and called. Will dive into that later as well

**Comment by juanmanso** [comment]
I'm a bit stuck with this task, gonna put it back to the ready column and take something else to get back to rhythm and come with a fresh mind to go at it again later 💪

**Comment by juanmanso** [comment]
Gonna do some pairing with @obasilakis here to move things forward 💪

**Comment by juanmanso** [comment]
Pushed the changes I'm most certain to be good, have some others stashed locally but at least have them pushed so its visible for everyone at:

- `feat/MR-590/upgrate-rainbowkit-wagmi-viem`

**Comment by juanmanso** [comment]
🟢 **[UPDATE]** _Got story to work with new provider_

| <video src='https://github.com/user-attachments/assets/45ef3093-7d12-4c28-aa0f-490c47882d6f'> |
| :--: |

**NEXT STEPS**
- Fix storybook server errors (mainly due to some key files not being updated)
- Filter networks through `SupportedNetworks` 
- Add Masa Network support
- Test other functionalities

Code is available here: [`e1a5f22`](https://github.com/masa-finance/masa-react/commit/e1a5f22687651311221967f07d67d4540a622b66)

**Comment by juanmanso** [comment]
🟢 **[UPDATE]** _Splitting work with @obasilakis_

Gonna split the work as such:

- @obasilakis is taking the syntax errors to be fixed
- @juanmanso is taking the masa mainnet and testnet addition to the chains array

**Comment by juanmanso** [comment]
Since this task no longer blocks [masa-finance/masa-next-sbt#1444](https://github.com/masa-finance/masa-next-sbt/issues/1444) due to the new findings and development cannot be finished in a couple of more hours, we dropped its priority and we moved it to the `Ready` column below the waterline

**Comment by juanmanso** [comment]
🟢 **[EOD UPDATE]**

Changes have been pushed to this branch:
- `feat/MR-590/update-rainbowkit-wagmi-viem`

### Added Masa and MasaTestnet

| _Demo_ |
| :--: |
| <video src='https://github.com/user-attachments/assets/6ce817d5-f6bd-40a0-af51-a25ed859a8c5'> |


### Syntax errors

| ![Image](https://github.com/user-attachments/assets/253473f3-e4ab-4d22-8bfb-a027ca000375) |
| :--: |

We are at 24 errors while running the storybook. Gonna continue fixing these next week 🚀
